src/System.Linq.Expressions/src/System/Linq/Expressions/BlockExpression.cs

```diff
@@ -937,5 +955,20 @@ internal static void ValidateVariables(ReadOnlyCollection<ParameterExpression> v
                 set.Add(v);
             }
         }
+
+        private static BlockExpression GetOptimizedBlockExpression(IReadOnlyList<Expression> expressions)
+        {
+            switch (expressions.Count)
+            {
+                case 2: return Block(expressions[0], expressions[1]);
+                case 3: return Block(expressions[0], expressions[1], expressions[2]);
+                case 4: return Block(expressions[0], expressions[1], expressions[2], expressions[3]);
+                case 5: return Block(expressions[0], expressions[1], expressions[2], expressions[3], expressions[4]);
+                default:
+                    ContractUtils.RequiresNotEmptyList(expressions, "expressions");
+                    RequiresCanRead(expressions, "expressions");
+                    return new BlockN(expressions.ToArray());
```

Comment:
Could `ToReadOnly()` serve better here? It likewise promises that changes to the source enumerable won't change the copy held in the `BlockN`, but won't create a new collection if `Expressions` is accessed (very likely) at the point of its access, and won't create a new collection or if the source enumerable was already a `ReadOnlyCollection`, as happens with the default `Visit` behaviour.